README
NAME
    Clone::PP - Recursively copy Perl datatypes

SYNOPSIS
      use Clone::PP qw(clone);
  
      $a = { 'foo' => 'bar', 'move' => 'zig' };
      $b = [ 'alpha', 'beta', 'gamma', 'vlissides' ];
      $c = new Foo();
  
      $d = clone($a);
      $e = clone($b);
      $f = clone($c);
  
      # or
  
      use Clone::PP;
      push @Foo::ISA, 'Clone';
  
      $a = new Foo;
      $b = $a->clone();

DESCRIPTION
    This module provides a clone() method which makes recursive copies of
    nested hash, array, scalar and reference types, including tied variables
    and objects.

    The clone() function takes a scalar argument and an optional parameter
    that can be used to limit the depth of the copy. To duplicate lists,
    arrays or hashes, pass them in by reference. e.g.

      my $copy = clone (\@array);
      # or
      my %copy = %{ clone (\%hash) };  

SEE ALSO
    For a faster implementation in XS, see the Clone manpage.

    For a slower, but more flexible solution use the dclone function from
    <Storable>.

CREDITS AND COPYRIGHT
    Developed by Matthew Simon Cavalletto, simonm@cavalletto.org. Mode
    modules from Evolution Softworks are available at www.evoscript.org.
    Copyright 2003 Matthew Simon Cavalletto.

    Interface based on Clone by Ray Finch, rdf@cpan.org. Portions Copyright
    2001 Ray Finch.

    Code based on initial design from Ref.pm. Portions Copyright 1994 David
    Muir Sharnoff.

    This module is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl module Clone::PP

1.06 2014-07-20 NEILB
    - Changed the copyright on dclone.t, dump.pl, and tied.pl in t/,
      to match the rest of the dist, "same as perl itself".
      These changes were approved by the copyright owner, Raphael Manfredi.

1.05 2014-02-17 NEILB
    - Added github repo to metadata (thanks dsteinbrunner)
    - Added github repo to doc

1.04 2014-02-17 NEILB
    - I had erroneously listed Test::Array as a test prereq,
      but the package is defined in 01array.t where it's used.
      RT#93082 - thanks to Dagfinn Ilmari Mannsaker.

1.03 2014-02-16 NEILB
    - Added Changes file
    - Fixed typos from dsteinbrunner in RT#86337
    - Added "use warnings" and in the process fixed RT#17121
    - Min perl version 5.6.0 in code and metadata
    - Added license type to metadata and renamed L&C section in pod
    - Expanded the SEE ALSO section
    - Added prereqs to PREREQ_PM in Makefile.PL
    - Added test prereqs in TEST_REQUIRES

1.02 2003-08-28

1.01 2003-08-27

1.00 2003-08-25
    - First release to CPAN


t/tied.pl
#!./perl

# $Id: tied.pl,v 0.11 2001/07/29 19:31:05 ray Exp $
#
#  Copyright (c) 1995-1998, Raphael Manfredi
#  
#  You may redistribute this file under the same terms as Perl 5 itself.
#
# $Log: tied.pl,v $
# Revision 0.11  2001/07/29 19:31:05  ray
# VERSION 0.11
#
# Revision 0.10  2001/04/29 21:56:10  ray
# VERSION 0.10
#
# Revision 0.9  2001/03/05 00:11:49  ray
# version 0.9
#
# Revision 0.9  2000/08/21 23:06:34  ray
# added support for code refs
#
# Revision 0.8  2000/08/11 17:08:36  ray
# Release 0.08.
#
# Revision 0.7  2000/08/01 00:43:48  ray
# release 0.07.
#
# Revision 0.6.2.1  2000/08/01 00:42:53  ray
# modified to use as a require statement.
#
# Revision 0.6  2000/08/01 01:38:38  ray
# "borrowed" code from Storable
#
# Revision 0.6  1998/06/04 16:08:40  ram
# Baseline for first beta release.
#

require 't/dump.pl';

package TIED_HASH;

sub TIEHASH {
	my $self = bless {}, shift;
	return $self;
}

sub FETCH {
	my $self = shift;
	my ($key) = @_;
	$main::hash_fetch++;
	return $self->{$key};
}

sub STORE {
	my $self = shift;
	my ($key, $value) = @_;
	$self->{$key} = $value;
}

sub FIRSTKEY {
	my $self = shift;
	scalar keys %{$self};
	return each %{$self};
}

sub NEXTKEY {
	my $self = shift;
	return each %{$self};
}

sub CLEAR {
	%$self = ();  
}

package TIED_ARRAY;

sub TIEARRAY {
	my $self = bless [], shift;
	return $self;
}

sub FETCH {
	my $self = shift;
	my ($idx) = @_;
	$main::array_fetch++;
	return $self->[$idx];
}

sub STORE {
	my $self = shift;
	my ($idx, $value) = @_;
	$self->[$idx] = $value;
}

sub FETCHSIZE {
	my $self = shift;
	return @{$self};
}

sub CLEAR {
	@$self = ();  
}

sub EXTEND  { }

package TIED_SCALAR;

sub TIESCALAR {
	my $scalar;
	my $self = bless \$scalar, shift;
	return $self;
}

sub FETCH {
	my $self = shift;
	$main::scalar_fetch++;
	return $$self;
}

sub STORE {
	my $self = shift;
	my ($value) = @_;
	$$self = $value;
}

sub CLEAR {
	$$self = ();  
}

1;


Makefile.PL
use ExtUtils::MakeMaker;

my $mm_ver = $ExtUtils::MakeMaker::VERSION;
if ($mm_ver =~ /_/) { # dev version
    $mm_ver = eval $mm_ver;
    die $@ if $@;
}

WriteMakefile(
    'NAME'	=> 'Clone::PP',
    'VERSION_FROM' => 'lib/Clone/PP.pm', 

    PREREQ_PM      => {
        'Exporter' => 0,
        'strict'   => 0,
        'warnings' => 0,
        'vars'     => 0,
    },

    ($mm_ver >= 6.48
        ? (MIN_PERL_VERSION => 5.006)
        : ()
    ),

    ($mm_ver >= 6.31
        ? (LICENSE => 'perl_5')
        : ()
    ),

    ($mm_ver >= 6.64
        ? (TEST_REQUIRES => {
               'Data::Dumper' => 0,
               'Benchmark'    => 0,
               'Carp'         => 0,
          })
        : ()
    ),

    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/neilbowers/Clone-PP.git',
                web  => 'https://github.com/neilbowers/Clone-PP',
            },
        },
    },

);



lib/Clone/PP.pm
package Clone::PP;

use 5.006;
use strict;
use warnings;
use vars qw($VERSION @EXPORT_OK);
use Exporter;

$VERSION = 1.05;

@EXPORT_OK = qw( clone );
sub import { goto &Exporter::import } # lazy Exporter

# These methods can be temporarily overridden to work with a given class.
use vars qw( $CloneSelfMethod $CloneInitMethod );
$CloneSelfMethod ||= 'clone_self';
$CloneInitMethod ||= 'clone_init';

# Used to detect looped networks and avoid infinite recursion. 
use vars qw( %CloneCache );

# Generic cloning function
sub clone {
  my $source = shift;

  return undef if not defined($source);
  
  # Optional depth limit: after a given number of levels, do shallow copy.
  my $depth = shift;
  return $source if ( defined $depth and $depth -- < 1 );
  
  # Maintain a shared cache during recursive calls, then clear it at the end.
  local %CloneCache = ( undef => undef ) unless ( exists $CloneCache{undef} );
  
  return $CloneCache{ $source } if ( defined $CloneCache{ $source } );
  
  # Non-reference values are copied shallowly
  my $ref_type = ref $source or return $source;
  
  # Extract both the structure type and the class name of referent
  my $class_name;
  if ( "$source" =~ /^\Q$ref_type\E\=([A-Z]+)\(0x[0-9a-f]+\)$/ ) {
    $class_name = $ref_type;
    $ref_type = $1;
    # Some objects would prefer to clone themselves; check for clone_self().
    return $CloneCache{ $source } = $source->$CloneSelfMethod() 
				  if $source->can($CloneSelfMethod);
  }
  
  # To make a copy:
  # - Prepare a reference to the same type of structure;
  # - Store it in the cache, to avoid looping if it refers to itself;
  # - Tie in to the same class as the original, if it was tied;
  # - Assign a value to the reference by cloning each item in the original;
  
  my $copy;
  if ($ref_type eq 'HASH') {
    $CloneCache{ $source } = $copy = {};
    if ( my $tied = tied( %$source ) ) { tie %$copy, ref $tied }
    %$copy = map { ! ref($_) ? $_ : clone($_, $depth) } %$source;
  } elsif ($ref_type eq 'ARRAY') {
    $CloneCache{ $source } = $copy = [];
    if ( my $tied = tied( @$source ) ) { tie @$copy, ref $tied }
    @$copy = map { ! ref($_) ? $_ : clone($_, $depth) } @$source;
  } elsif ($ref_type eq 'REF' or $ref_type eq 'SCALAR') {
    $CloneCache{ $source } = $copy = \( my $var = "" );
    if ( my $tied = tied( $$source ) ) { tie $$copy, ref $tied }
    $$copy = clone($$source, $depth);
  } else {
    # Shallow copy anything else; this handles a reference to code, glob, regex
    $CloneCache{ $source } = $copy = $source;
  }
  
  # - Bless it into the same class as the original, if it was blessed;
  # - If it has a post-cloning initialization method, call it.
  if ( $class_name ) {
    bless $copy, $class_name;
    $copy->$CloneInitMethod() if $copy->can($CloneInitMethod);
  }
  
  return $copy;
}

1;

__END__

=head1 NAME

Clone::PP - Recursively copy Perl datatypes

=head1 SYNOPSIS

  use Clone::PP qw(clone);
  
  $item = { 'foo' => 'bar', 'move' => [ 'zig', 'zag' ]  };
  $copy = clone( $item );

  $item = [ 'alpha', 'beta', { 'gamma' => 'vlissides' } ];
  $copy = clone( $item );

  $item = Foo->new();
  $copy = clone( $item );

Or as an object method:

  require Clone::PP;
  push @Foo::ISA, 'Clone::PP';
  
  $item = Foo->new();
  $copy = $item->clone();

=head1 DESCRIPTION

This module provides a general-purpose clone function to make deep
copies of Perl data structures. It calls itself recursively to copy
nested hash, array, scalar and reference types, including tied
variables and objects.

The clone() function takes a scalar argument to copy. To duplicate
arrays or hashes, pass them in by reference:

  my $copy = clone(\@array);    my @copy = @{ clone(\@array) };
  my $copy = clone(\%hash);     my %copy = %{ clone(\%hash) };

The clone() function also accepts an optional second parameter that
can be used to limit the depth of the copy. If you pass a limit of
0, clone will return the same value you supplied; for a limit of
1, a shallow copy is constructed; for a limit of 2, two layers of
copying are done, and so on.

  my $shallow_copy = clone( $item, 1 );

To allow objects to intervene in the way they are copied, the
clone() function checks for a couple of optional methods. If an
object provides a method named C<clone_self>, it is called and the
result returned without further processing. Alternately, if an
object provides a method named C<clone_init>, it is called on the
copied object before it is returned.

=head1 BUGS

Some data types, such as globs, regexes, and code refs, are always copied shallowly.

References to hash elements are not properly duplicated. (This is why two tests in t/dclone.t that are marked "todo".) For example, the following test should succeed but does not:

  my $hash = { foo => 1 }; 
  $hash->{bar} = \{ $hash->{foo} }; 
  my $copy = clone( \%hash ); 
  $hash->{foo} = 2; 
  $copy->{foo} = 2; 
  ok( $hash->{bar} == $copy->{bar} );

To report bugs via the CPAN web tracking system, go to 
C<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Clone-PP> or send mail 
to C<Dist=Clone-PP#rt.cpan.org>, replacing C<#> with C<@>.

=head1 SEE ALSO

L<Clone> - a baseclass which provides a C<clone()> method.

L<MooseX::Clone> - find-grained cloning for Moose objects.

The C<dclone()> function in L<Storable>.

L<Data::Clone> -
polymorphic data cloning (see its documentation for what that means).

L<Clone::Any> - use whichever of the cloning methods is available.

=head1 REPOSITORY

L<https://github.com/neilbowers/Clone-PP>

=head1 AUTHOR AND CREDITS

Developed by Matthew Simon Cavalletto at Evolution Softworks. 
More free Perl software is available at C<www.evoscript.org>.


=head1 COPYRIGHT AND LICENSE

Copyright 2003 Matthew Simon Cavalletto. You may contact the author
directly at C<evo@cpan.org> or C<simonm@cavalletto.org>.

Code initially derived from Ref.pm. Portions Copyright 1994 David Muir Sharnoff.

Interface based by Clone by Ray Finch with contributions from chocolateboy.
Portions Copyright 2001 Ray Finch. Portions Copyright 2001 chocolateboy. 

You may use, modify, and distribute this software under the same terms as Perl.

=cut
